### PR TITLE
[OO] Add maximum initial quantity to ortools

### PR DESCRIPTION
**Pull request by braktar**
Requirement
- [x] https://github.com/Mapotempo/optimizer-ortools/pull/22

### CODE REVIEW

File: wrappers/ortools.rb [hunk 1]
```diff
@@ -100,6 +99,17 @@ def solve(vrp, job, thread_proc = nil, &block)
           empty: false
         }
       }
+      sum_quantities = vrp.units.map{ |unit| [unit.id, 0] }.to_h
+      vrp.services.each{ |service|
+        service.quantities.each{ |q|
+          sum_quantities[q.unit.id] += (q.value || 0).abs
+        }
+      }
+      vrp.shipments.each{ |shipment|
```
Comment (fonsecadeline) [inline wrappers/ortools.rb:108]: I guess we can remove this part now ?

File: wrappers/ortools.rb [hunk 2]
```diff
@@ -366,7 +376,8 @@ def solve(vrp, job, thread_proc = nil, &block)
             OrtoolsVrp::Capacity.new(
               limit: (q&.limit && q.limit < 1e+22) ? q.limit : -1,
               overload_multiplier: q&.overload_multiplier || 0,
-              counting: unit&.counting || false
+              counting: unit&.counting || false,
+              initial: q.initial || [sum_quantities[unit.id], q.limit].compact.max
```
Comment (senhalil) [inline wrappers/ortools.rb:380]: I suspect this `max` should be `min`, right?


Can there be a reason to take the max -- i.e., why would we load the vehicle more than the total_demand if the capacity.limit > total_demand?

Similarly if the the vehicle capacity is less than total_demand, then initial load is set to total_demand > capacity.limit.

File: wrappers/ortools.rb [hunk 3]
```diff
@@ -100,6 +99,17 @@ def solve(vrp, job, thread_proc = nil, &block)
           empty: false
         }
       }
+      sum_quantities = vrp.units.map{ |unit| [unit.id, 0] }.to_h
+      vrp.services.each{ |service|
+        service.quantities.each{ |q|
+          sum_quantities[q.unit.id] += (q.value || 0).abs
+        }
+      }
```
Comment (senhalil) [inline wrappers/ortools.rb:107]: This double counts the demand of the services in shipment relation. 

I think you can first do a loop on `vrp.relations.select{:shipment}.each.linked_services.last.quantities.each`
then in the `vrp.services.each` loop skip the `service.relations.any?{:shipment}`

Comment (senhalil) [reply wrappers/ortools.rb:107]: I think `total_quantities` or `quantity_totals` is easier to understand than `sum_quantities`.

File: wrappers/wrapper.rb [hunk 4]
```diff
@@ -56,7 +56,7 @@ def assert_vehicles_start_or_end(vrp)
 
     def assert_vehicles_no_capacity_initial(vrp)
       vrp.vehicles.none?{ |vehicle|
-        vehicle.capacities.find{ |c| c.initial&.positive? }
+        vehicle.capacities.any?{ |c| c.initial && c.initial < c.limit }
```
Comment (senhalil) [inline wrappers/wrapper.rb:59]: I don't quite get this condition.

Comment (senhalil) [reply wrappers/wrapper.rb:59]: Does `initial_load` signify the existing load in the vehicle? If so, how come it can be greater than the vehicle capacity?

**Review by senhalil — CHANGES_REQUESTED** [review]

File: wrappers/ortools.rb (hunk 2 above)
Comment (braktar) [reply wrappers/ortools.rb:380]: if we have an overload_multiplier, we can allow the sum of the quantities loaded to exceed the capacity, the `max` is used at this purpose

File: wrappers/wrapper.rb (hunk 4 above)
Comment (braktar) [reply wrappers/wrapper.rb:59]: If the initial capacity is greater than the capacity, it supposes or that we have a overload multiplier, or the initial capacity is not restricted

Comment (senhalil) [reply wrappers/wrapper.rb:59]: My question was due to the definition miss-match :) initial-load vs initial-capacity. it makes sense now

**Review by fonsecadeline — APPROVED** [review]

**Pull request merged**